### DISCUSSION TITLE
Value & revision for a transaction entry

### DISCUSSION THREAD

**Discussion by grossvater** [codenotary/immudb#1261] [Q&A] [answered]
Hi,
I loop over transactions with transaction reader (for an embedded store):

```
rtx := st.NewTxHolder()
r, err := st.NewTxReader(1, false, rtx)

for {
    txi, err := r.Read()
    for _, e := range txi.Entries() {
        key := e.Key()
        // value := ??
        // revision := ??
    }
}
```

How could I get the revision and value set by the transaction?
Thanks.

**Comment by jeroiraz** [comment, marked as answer]
When iterating over raw transactions, entries do not hold the actual payload but a reference to the value (and its digest). The store provides the `ReadValue` method to fetch the actual value:

```go
value, err := st.ReadValue(e)
```

**Reply by grossvater** [reply]
Great! What about the revision? I would have expected to be part of the entry.

**Reply by jeroiraz** [reply]
revision is not part of the raw transaction data, at least not in current implementation. The reason for this is that keeping track of the revision requires the indexing to be fully synchronized or some other data structure, which may not be as fast as the commit log.

revision of an entry requires reading the value by its associated key `st.Get(key)`

**Reply by grossvater** [reply]
I understand. But still, if need the revision of key, created by a specific transaction, how could I get it? I saw

`History(key []byte, offset uint64, descOrder bool, limit int) (txs []uint64, hCount uint64, err error)`

so taking the index of the transaction from `txs` is the way?

**Reply by jeroiraz** [reply]
hCount gives the total number of revisions (or key updates).
If desOrder is true:
 - the txs[0] is the latest update, so revision would match hCount
- then revision would be hCount - txIndex
If not descending order:
-  then the first update is at txs[0]
- then revision would be 1 + txIndex

Feel free to open a feature request on how to get the revision for a key at a specific tx

**Reply by grossvater** [reply]
Ok, thanks for all the information.
